pkg/aws/ec2/api/helper.go

```diff
@@ -434,77 +439,120 @@ func (h *ec2APIHelper) GetInstanceDetails(instanceId *string) (*ec2.Instance, er
+			// Describe the network interface on which the new IP or prefixes are assigned
 			interfaces, err := h.DescribeNetworkInterfaces([]*string{&eniID})
-			// Re initialize the slice so we don't add IPs multiple time
-			assignedIPs = []string{}
-			if err == nil && len(interfaces) == 1 && interfaces[0].PrivateIpAddresses != nil {
-				// Get the map of IPs returned by the describe network interface call
-				ipAddress := map[string]bool{}
-				for _, ipAddr := range interfaces[0].PrivateIpAddresses {
-					ipAddress[*ipAddr.PrivateIpAddress] = true
-				}
-				// Verify the describe network interface returns all the IPs that were assigned in the
-				// AssignPrivateIPAddresses call
-				for _, ip := range assignPrivateIPOutput.AssignedPrivateIpAddresses {
-					if _, ok := ipAddress[*ip.PrivateIpAddress]; !ok {
-						// Even if one IP is not assigned, set the error so that we only return only the IPs that
-						// are successfully assigned on the ENI
-						err = ErrIPNotAttachedYet
-					} else {
-						assignedIPs = append(assignedIPs, *ip.PrivateIpAddress)
+			// Re-initialize the slice so that we don't add IP resources multiple times
+			assignedResources = []string{}
+
+			if err == nil && len(interfaces) == 1 {
+				if resourceType == config.ResourceTypeIPv4Address && interfaces[0].PrivateIpAddresses != nil {
+					// Get the map of IPs returned by the describe network interface call
+					ipAddress := map[string]bool{}
+					for _, ipAddr := range interfaces[0].PrivateIpAddresses {
+						ipAddress[*ipAddr.PrivateIpAddress] = true
+					}
+					// Verify describe network interface returns all the IPs that were assigned in the
+					// AssignPrivateIPAddresses call
+					for _, ip := range assignPrivateIPOutput.AssignedPrivateIpAddresses {
+						if _, ok := ipAddress[*ip.PrivateIpAddress]; !ok {
+							// Even if one IP is not assigned, set the error so that we only return only the IPs that
+							// are successfully assigned on the ENI
+							err = ErrIPResourceNotAttachedYet
+						} else {
+							assignedResources = append(assignedResources, *ip.PrivateIpAddress)
+						}
+					}
+					return err
+				} else {
+					if resourceType == config.ResourceTypeIPv4Prefix && interfaces[0].Ipv4Prefixes != nil {
+						// Get the map of IP prefixes returned by the describe network interface call
+						ipPrefixes := map[string]bool{}
+						for _, ipPrefix := range interfaces[0].Ipv4Prefixes {
+							ipPrefixes[*ipPrefix.Ipv4Prefix] = true
+						}
+						// Verify describe network interface returns all the IP prefixes that were assigned in the
+						// AssignPrivateIPAddresses call
+						for _, prefix := range assignPrivateIPOutput.AssignedIpv4Prefixes {
+							if _, ok := ipPrefixes[*prefix.Ipv4Prefix]; !ok {
+								// Even if one prefix is not assigned, set the error so that we only return the IP prefixes that
+								// are successfully assigned on the ENI
+								err = ErrIPResourceNotAttachedYet
```

Comment:
Can we explicitly use `ErrIPResourceNotAttachedYet` and `ErrPrefixResourceNotAttachedYet`, and then in the retryable check func we can check if either one of them and log the error type? This may help debugging.

Reply:
Added both errors explicitly. Thanks for the suggestion.